### PR TITLE
[luci] Introduce FoldNegPass

### PR DESCRIPTION
**Pull request by llFreetimell**
Parent Issue : #7147

This commit introduces `FoldNegPass` in `luci`.

ONE-DCO-1.0-Signed-off-by: Seok NamKoong <seok9311@naver.com>

### CODE REVIEW

File: compiler/luci/pass/src/FoldNegPass.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ */
+
+#include "luci/Pass/FoldNegPass.h"
+
+#include <luci/IR/CircleNodes.h>
+
+namespace
+{
+
+bool fold_neg(luci::CircleNeg *neg)
```
Comment (seanshpark) [inline compiler/luci/pass/src/FoldNegPass.cpp:24]: Could you please add a diagram explaining what this pass does?

File: compiler/luci/pass/src/FoldNegPass.test.cpp [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+    _neg->x(_x);
+
+    _neg->name("neg");
+    _x->name("x");
+  }
+
+  loco::Node *createFoldedPattern() override { return _neg; }
+
+  void set_qparam()
```
Comment (seanshpark) [inline compiler/luci/pass/src/FoldNegPass.test.cpp:53]: Why is this needed?

File: compiler/luci/pass/src/FoldNegPass.cpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      case loco::DataType::FLOAT32:
+        for (uint32_t i = 0; i < const_x->size<loco::DataType::FLOAT32>(); ++i)
+          const_x->at<loco::DataType::FLOAT32>(i) *= -1.0;
+        break;
+      default:
+        return false;
+    }
+  }
+
+  loco::replace(neg).with(const_x);
```
Comment (seanshpark) [inline compiler/luci/pass/src/FoldNegPass.cpp:50]: if `const_x` is used as input to another node, this will damage the graph.
usuallly we leave the `const_x` node, create a clone of it and set the values and use it.

File: compiler/luci/pass/src/FoldNegPass.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  // Check neg has const input
+  auto const_x = dynamic_cast<luci::CircleConst *>(neg->x());
+  if (not const_x)
+    return false;
+
+  // If quantparam exists, fold by scale
+  if (auto qparam = const_x->quantparam())
+  {
+    for (uint32_t i = 0; i < qparam->scale.size(); ++i)
+      qparam->scale.at(i) *= -1.0;
```
Comment (seanshpark) [inline compiler/luci/pass/src/FoldNegPass.cpp:35]: there are some models that have `quantparam` by tflite converter, but set with meaningless values.
for that case, does it make correct folding?

**Comment by seanshpark** [comment]
@llFreetimell , please prepare a draft PR and do this
(1) add a recipe with the network and add/run `circle2circle-dredd-recipe-test`
(2) and this is optional, add/run `luci-pass-value-test`: I'm not not sure of `quantparam` change with _luci-interpreter_

File: compiler/luci/pass/src/FoldNegPass.cpp (hunk 3 above)
Comment (llFreetimell) [reply compiler/luci/pass/src/FoldNegPass.cpp:50]: Oh, I missed this, thanks :)

File: compiler/luci/pass/src/FoldNegPass.cpp (hunk 4 above)
Comment (llFreetimell) [reply compiler/luci/pass/src/FoldNegPass.cpp:35]: If `CircleConst` has `quantparam`, following equation can be calculated.

```
f = (q - zeropoint) * scale
-f = (q - zeropoint) * (-scale)
```

So I think it is not related with the meaningless values.

**Comment by llFreetimell** [comment]
> please prepare a draft PR

I thought that #7147 is really simple and this is almost all of to-do lists but I will do it.


> (1) add a recipe with the network and add/run circle2circle-dredd-recipe-test
> (2) and this is optional, but highly recommended, add/run luci-pass-value-test

I am not object to these but just curious, why most of other passes did not have this test until now...?

File: compiler/luci/pass/src/FoldNegPass.cpp (hunk 4 above)
Comment (jinevening) [inline compiler/luci/pass/src/FoldNegPass.cpp:35]: Negative scale seems risky to me. In the normally quantized model, zp will be changed after Neg is applied.

Do you have to fold Neg for quantized model? If not, IMHO it would be better to support only FP32 in this pass.

Comment (seanshpark) [reply compiler/luci/pass/src/FoldNegPass.cpp:35]: @jinevening pointed out what I was worried in mind. 
This code doesn't check the dtype so would also run with Float32. 
What I wanted to point out is that float models also have `quantparam`, is this what you have in mind too?

**Comment by seanshpark** [comment]
> I am not object to these but just curious, why most of other passes did not have this test until now...?

I think they are simple or easly predictable or just lazy not to make it.
What made me ask of the test is about the `quantparam` usage which is not sure for me.

**Comment by llFreetimell** [comment]
Let's reopen this when it is needed
